File: minRefuelStops.java
import java.util.PriorityQueue;
import java.util.List;

class Solution {
    public int minRefuelStops(int target, int tank, List<int[]> stations) {
        int ans = 0, prev = 0;

        PriorityQueue<Integer> pq = new PriorityQueue<>((a, b) -> b - a);

        for (int[] station : stations) {
            int location = station[0];
            int fuel = station[1];

            int d = location - prev;

            tank -= d;

            while (!pq.isEmpty() && tank < 0) {
                tank += pq.poll();
                ans++;
            }

            if (tank < 0) return -1;

            prev = location;
            pq.offer(fuel);
        }

        tank -= (target - prev);

        while (!pq.isEmpty() && tank < 0) {
            tank += pq.poll();
            ans++;
        }

        if (tank < 0) return -1;

        return ans;
    }
}

File: minRefuelStops.py
import heapq

class Solution:
    def minRefuelStops(self, target, tank, stations):
        ans, prev = 0, 0
        pq = []

        for station in stations:
            location, fuel = station[0], station[1]
            d = location - prev
            tank -= d

            while pq and tank < 0:
                tank += -heapq.heappop(pq)
                ans += 1

            if tank < 0:
                return -1

            prev = location
            heapq.heappush(pq, -fuel)

        tank -= (target - prev)

        while pq and tank < 0:
            tank += -heapq.heappop(pq)
            ans += 1

        if tank < 0:
            return -1

        return ans
